server/src/main/java/org/elasticsearch/health/HealthPeriodicLogger.java

```diff
@@ -246,10 +264,35 @@ public void clusterChanged(ClusterChangedEvent event) {
     }
 
     @Override
-    public void close() {
-        SchedulerEngine engine = scheduler.get();
-        if (engine != null) {
-            engine.stop();
+    protected void doStart() {
+        logger.debug("Periodic health logger is starting.");
+    }
+
+    /**
+     * Stopping means that the periodic health logger will not schedule any more runs. If the logger is currently running it will
+     * let this run finish, but it will cancel any future scheduling, and it will deregister the cluster state listener.
+     */
+    @Override
+    protected void doStop() {
+        clusterService.removeListener(this);
+        logger.debug("Periodic health logger is stopping.");
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+        logger.debug("Periodic health logger is closing.");
+        try {
+            // The health API is expected to be a quick call, so we do not need a very long timeout
+            if (currentlyRunning.tryAcquire(2, TimeUnit.SECONDS)) {
+                currentlyRunning.release();
```

Comment:
If we are closed, should we just not release the semaphore? That would prevent this from being restarted ever (which may be more correct), otherwise it seems strange that we don't actually prevent running once the service is closed (I know we removed the listener, but if there were a problem with unregistering it we'd keep running even though the service had been closed). Normally it seems like we'd do this with a `running` AtomicBoolean, but we don't have this here.

Reply:
I felt a bit uncomfortable having a lock acquired and never released, but considering your explanation it makes sense. I will remove the release.